### ISSUE TITLE
Static cache of `File` nodes passed to remark plugins

### ISSUE THREAD

**Issue by stevetweeddale** [ChristopherBiscardi/gatsby-mdx#257]
Remark plugins get passed an array of `File` nodes, so they can hook up file references in the markdown to those files on disk and do special things (eg. `gatsby-remark-images` and `gatsby-remark-copy-linked-files`). When doing a production build this array of `File` nodes gets statically cached, globally, in [`get-source-plugins-as-remark-plugins.js`](https://github.com/ChristopherBiscardi/gatsby-mdx/blob/edafe526c5a9d124f4f8db68932967f3a2377b66/packages/gatsby-mdx/utils/get-source-plugins-as-remark-plugins.js#L40-L42) (note the variable is declared at the top level of the file)

What this means is that, if you have multiple instances of `gatsby-source-filesystem` generating markdown `File` nodes (or in my actual use case, another plugin that creates them) then subsequent runs of the MDX transformer will pass a stale array of `File` nodes to the remark plugins (since other `File` nodes may have just been created).

I've spun up a [minimal repro branch](https://github.com/stevetweeddale/gatsby-mdx-minimal-repro/tree/remark-plugin-race-condition) that uses `gatsby-remark-copy-linked-files` to demonstrate the issue. 

1. Run `gatsby build && gatsby serve`
2. Compare '/page' (generated from the first instance) with '/second/' (second instance)

Both pages should have an image tag generated by `gatsby-remark-copy-linked-files`,  but as it is, the second page is unprocessed by `gatsby-remark-copy-linked-files`, because the array of `File`'s it was passed didn't include the new image referenced on that page.

I'm now using a nasty patch hamstringing the conditional, so that `getNodes()` gets called every time, but I'm aware the caching was probably put there for a reason! If we were going to maintain any caching I'd assume it would have to be at a higher level and somehow flushed when new `File` nodes were added to the graph. I see that a cache object gets passed to `getSourcePluginsAsRemarkPlugins`, but I've not really explored higher up the call stack so would need some significant pointers if I was going to take a look at it!

**Comment by danoc** [comment]
Ran into this issue in #258.

Thanks for the debugging and workaround idea, @stevetweeddale. 

For those interested, this is how I'm implementing the workaround in my `build` script within `package.json`:

```bash
npx replace-in-file 'process.env.NODE_ENV !== `production` || !fileNodes' 'true || process.env.NODE_ENV !== `production` || !fileNodes' ../node_modules/gatsby-mdx/utils/get-source-plugins-as-remark-plugins.js && gatsby build
```

(Tried using `sed` but couldn't make it easily work on both OS X and Netlify's environment.)

I'd recommend also pinning the `gatsby-mdx` since this workaround is quite specific to that version.

**Comment by ChristopherBiscardi** [comment]
I published this as 0.3.5.

Thanks for the repro repo @stevetweeddale . It was very useful to test and resolve this.
